File: dsa master/linked list/InsertNdelete.java
class Node {

    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

public class LinkedListOperations {

    Node head;

    // ----------------------------
    // Insert at Beginning
    // ----------------------------
    void insertAtBeginning(int data) {

        Node newNode = new Node(data);

        newNode.next = head;

        head = newNode;
    }

    // ----------------------------
    // Insert at End
    // ----------------------------
    void insertAtEnd(int data) {

        Node newNode = new Node(data);

        if (head == null) {
            head = newNode;
            return;
        }

        Node current = head;

        while (current.next != null) {
            current = current.next;
        }

        current.next = newNode;
    }

    // ----------------------------
    // Insert at Position
    // Position starts from 0
    // ----------------------------
    void insertAtPosition(int position, int data) {

        if (position == 0) {
            insertAtBeginning(data);
            return;
        }

        Node newNode = new Node(data);

        Node current = head;

        for (int i = 0; i < position - 1 && current != null; i++) {
            current = current.next;
        }

        if (current == null) {
            System.out.println("Invalid Position");
            return;
        }

        newNode.next = current.next;

        current.next = newNode;
    }

    // ----------------------------
    // Delete from Beginning
    // ----------------------------
    void deleteBeginning() {

        if (head == null)
            return;

        head = head.next;
    }

    // ----------------------------
    // Delete from End
    // ----------------------------
    void deleteEnd() {

        if (head == null)
            return;

        if (head.next == null) {
            head = null;
            return;
        }

        Node current = head;

        while (current.next.next != null) {
            current = current.next;
        }

        current.next = null;
    }

    // ----------------------------
    // Delete at Position
    // ----------------------------
    void deleteAtPosition(int position) {

        if (head == null)
            return;

        if (position == 0) {
            head = head.next;
            return;
        }

        Node current = head;

        for (int i = 0; i < position - 1 && current.next != null; i++) {
            current = current.next;
        }

        if (current.next == null) {
            System.out.println("Invalid Position");
            return;
        }

        current.next = current.next.next;
    }

    // ----------------------------
    // Print List
    // ----------------------------
    void printList() {

        Node current = head;

        while (current != null) {

            System.out.print(current.data + " ");

            current = current.next;
        }

        System.out.println();
    }

    // ----------------------------
    // Main Method
    // ----------------------------
    public static void main(String[] args) {

        LinkedListOperations list = new LinkedListOperations();

        list.insertAtEnd(10);
        list.insertAtEnd(20);
        list.insertAtEnd(30);

        System.out.println("Original List");
        list.printList();

        list.insertAtBeginning(5);

        System.out.println("After Insert Beginning");
        list.printList();

        list.insertAtEnd(40);

        System.out.println("After Insert End");
        list.printList();

        list.insertAtPosition(2, 15);

        System.out.println("After Insert Position");
        list.printList();

        list.deleteBeginning();

        System.out.println("After Delete Beginning");
        list.printList();

        list.deleteEnd();

        System.out.println("After Delete End");
        list.printList();

        list.deleteAtPosition(2);

        System.out.println("After Delete Position");
        list.printList();
    }
}

File: dsa master/linked list/InsertNdelete.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


# ---------------- INSERTION ----------------

def insert_beginning(head, data):
    new_node = Node(data)

    new_node.next = head
    head = new_node

    return head


def insert_end(head, data):
    new_node = Node(data)

    if head is None:
        return new_node

    current = head

    while current.next is not None:
        current = current.next

    current.next = new_node

    return head


def insert_at_position(head, data, position):
    if position == 0:
        return insert_beginning(head, data)

    new_node = Node(data)

    current = head

    for _ in range(position - 1):
        if current is None:
            return head

        current = current.next

    if current is None:
        return head

    new_node.next = current.next
    current.next = new_node

    return head


# ---------------- DELETION ----------------

def delete_beginning(head):
    if head is None:
        return None

    return head.next


def delete_end(head):
    if head is None:
        return None

    # Only one node
    if head.next is None:
        return None

    current = head

    # Reach second-last node
    while current.next.next is not None:
        current = current.next

    current.next = None

    return head


def delete_at_position(head, position):
    if head is None:
        return None

    # Delete first node
    if position == 0:
        return head.next

    current = head

    # Reach node before the target
    for _ in range(position - 1):
        if current.next is None:
            return head

        current = current.next

    # Position doesn't exist
    if current.next is None:
        return head

    # Skip the target node
    current.next = current.next.next

    return head


# ---------------- TRAVERSAL ----------------

def traverse(head):
    current = head

    while current is not None:
        print(current.data, end=" → ")
        current = current.next

    print("None")


# ---------------- MAIN ----------------

def main():

    # Create linked list
    head = Node(10)
    head.next = Node(20)
    head.next.next = Node(30)

    print("Original list:")
    traverse(head)


    # Insert at beginning
    head = insert_beginning(head, 5)

    print("\nAfter inserting 5 at beginning:")
    traverse(head)


    # Insert at end
    head = insert_end(head, 40)

    print("\nAfter inserting 40 at end:")
    traverse(head)


    # Insert at position
    head = insert_at_position(head, 25, 3)

    print("\nAfter inserting 25 at position 3:")
    traverse(head)


    # Delete beginning
    head = delete_beginning(head)

    print("\nAfter deleting beginning:")
    traverse(head)


    # Delete end
    head = delete_end(head)

    print("\nAfter deleting end:")
    traverse(head)


    # Delete at position
    head = delete_at_position(head, 2)

    print("\nAfter deleting position 2:")
    traverse(head)


main()

        System.out.println("After Delete Position");
        list.printList();
    }
}
